### ISSUE TITLE
Multiple interfaces?

### ISSUE THREAD

**Issue by phated** [sunli829/async-graphql#18]
First, I wanted to say thank you for this amazing project! 🙇

I'm currently implementing a Relay client against my async-graphql server and they require everything to implement the `Node { id }` interface for every object.

If I implement `Node` for every interface, I'm not sure how to implement another interface on the same object. Have you considered this for the `#[async_graphql::Interface]` macro?

**Comment by sunli829** [comment]
Very easy，here's an example, `MyObj` implemented`IntefaceA` and `InterfaceB`.

```rust
struct MyObj;

#[async_graphql::Object]
impl MyObj {
    async fn value_a(&self) -> i32 {
        unimplemented!()
    }

    async fn value_b(&self) -> i32 {
        unimplemented!()
    }
}

#[async_graphql::Interface(field(name = "value_a", type = "i32"))]
struct InterfaceA(MyObj);

#[async_graphql::Interface(field(name = "value_b", type = "i32"))]
struct InterfaceB(MyObj);
```

**Comment by phated** [comment]
I think there might be a problem in representing that structure in a resolver.

If you implement:
```rust
pub struct QueryRoot;

#[async_graphql::Object]
impl QueryRoot {
  #[field]
  async fn my_objs(&self, _ctx: &Context<'_>) -> InterfaceA {
    (MyObj {}).into()
  }
}
```

In the above example, the server doesn't know about `InterfaceB` and the query execution fails.

I put together a sample repository at https://github.com/phated/async-graphql-interfaces that identifies the problem.

**Comment by phated** [comment]
I've been thinking *a lot* about Interfaces in GraphQL to support Relay in my data structures and I wonder if they are better represented in Rust by implementing a trait.

What do you think of this API?

```rust
struct MyObj;

#[async_graphql::Object]
impl MyObj {
  async fn value_c(&self) -> i32 {
    3
  }
}

// Would the macro go on trait or impl?
#[async_graphql::Interface]
trait InterfaceA {
  async fn value_a(&self) -> i32;
}

// Would the macro go on trait or impl?
#[async_graphql::Interface]
trait InterfaceB {
  async fn value_b(&self) -> i32;
}

// Maybe it would go on both?
#[async_graphql::Interface]
impl InterfaceA for MyObj {
  #[field]
  async fn value_a(&self) -> i32 {
    1
  }
}

// Maybe it would go on both?
#[async_graphql::Interface]
impl InterfaceB for MyObj {
  #[field]
  async fn value_b(&self) -> i32 {
    2
  }
}

pub struct QueryRoot;

#[async_graphql::Object]
impl QueryRoot {
  #[field]
  async fn my_objs(&self, _ctx: &Context<'_>) -> impl InterfaceA + impl InterfaceB {
    (MyObj {}).into()
  }
}
```

I'm not sure if this is possible with Rust macros because I don't understand them enough.

**Comment by sunli829** [comment]
This seems like a pretty good design, but unfortunately it cannot be implemented due to the limitations of the rust procedural macros. The current interface is defined in a way that I've thought about for a long time, is easy to use and easy to understand, and more importantly is rustfmt friendly. See how juniper defines the interface. It's similar to what you want, but not easy to use.

**Comment by sunli829** [comment]
I see what you mean. You want to return a combination of two interfaces. Let me think for a moment.

**Comment by phated** [comment]
Thank you, I was just going to write that. I believe it is a requirement by [GraphQL spec](https://spec.graphql.org/draft/#sec-Interfaces).

This is a common pattern when using Relay because their client adds the Node fragment onto every query.

Juniper doesn't support this as far as I can tell, which was the reason I found your awesome library.

**Comment by sunli829** [comment]
Is it something like that?

```graphql
interface Node {
  id: ID!
}

interface Resource implements Node {
  id: ID!
  url: String
}

interface Image implements Resource & Node {
  id: ID!
  url: String
  thumbnail: String
}
```

**Comment by sunli829** [comment]
Why not just return `MyObj`?
```rust
#[async_graphql::Object]
impl QueryRoot {
  #[field]
  async fn my_objs(&self, _ctx: &Context<'_>) -> MyObj{
    MyObj {}
  }
}
```

**Comment by phated** [comment]
Each interface doesn't need to implement the others in a chain. See https://spec.graphql.org/draft/#example-ab5e5

**Comment by phated** [comment]
MyObj isn't the only struct implemention the interface, I just trying to shrink the example.

**Comment by sunli829** [comment]
Does GraphQL support such a field type?
```rust
The type of the Query {
    myObj: InterfaceA + InterfaceB
}
```

**Comment by phated** [comment]
I believe each would need to be an object that implements each interface. The interface values seem to be attached to the type declaration.

**Comment by sunli829** [comment]
'Async-graphql' currently does not support interface implementing interface.
Is the following function what you want?
`MyObj` implements `InterfaceB`, and because `InterfaceB` implements `InterfaceA`, `MyObj` automatically implements `InterfaceA`.

```rust
    struct MyObj;

    #[async_graphql::Object]
    impl MyObj {
        #[field]
        async fn value_a(&self) -> i32 {
            1
        }

        #[field]
        async fn value_b(&self) -> i32 {
            2
        }

        #[field]
        async fn value_c(&self) -> i32 {
            3
        }
    }

    #[async_graphql::Interface(field(name = "value_a", type = "i32"))]
    struct InterfaceA;

    #[async_graphql::Interface(field(name = "value_b", type = "i32"), implements = "InterfaceA")]
    struct InterfaceB(MyObj);

    struct Query;

    #[Object]
    impl Query {
        #[field]
        async fn my_obj(&self) -> InterfaceB {
            MyObj.into()
        }
    }

    let schema = Schema::new(Query, EmptyMutation, EmptySubscription);
    let query = format!(
        r#"{{
            myObj {{
               ... on InterfaceA {{
                valueA
              }}
              ... on InterfaceB {{
                valueB
              }}
              ... on MyObj {{
                valueC
              }}
            }}
        }}"#
    );
    assert_eq!(
        schema.execute(&query).await.unwrap().data,
        serde_json::json!({
            "myObj": {
                "valueA": 1,
                "valueB": 2,
                "valueC": 3,
            }
        })
    );
```

**Comment by phated** [comment]
I believe that would solve my current problem, yes.

**Comment by sunli829** [comment]
This is really a big problem, and I'm going to solve it now. Thank you very much.😁

**Comment by sunli829** [comment]
Upgrade to `1.9.4` and take a look at this test.
https://github.com/sunli829/async-graphql/blob/fd156500e316e596a5ae47b40f6d99825b5990f7/tests/interface.rs#L4

**Comment by phated** [comment]
It seems to be working! Thank you so much 🙇 

I will keep testing it on my project and file anything I find.

**Comment by sunli829** [comment]
The last time the solution is not good, I think because I'm not support `http://spec.graphql.org/draft/#sec-Interfaces.Interfaces-Implementing-Interfaces` **(Draft)**, but in fact is another bug.
I fixed this bug in version `1.9.8` and removed the `implements` attribute on the `async_graphql::Interface`.

----------------------------

Below is the updated test, which requires manual registration because `InterfaceA` is not directly referenced by the `Schema`.
https://github.com/sunli829/async-graphql/blob/73aef368dfe2eb046f6d7c012f2576e19a002724/tests/interface.rs#L98
